### PR TITLE
Bug 1381840 - Schedule 1-Day Firefox Retention

### PR DESCRIPTION
**Pull request by acmiyaguchi**
This schedules the job using a slack of 4 days and 13 bits for the HLL
object. This corresponds to the observation of 99% of client activity
with roughly a 1% standard error introduced HLL.

This job shares components in mozetl and in telemetry-batch-view, so
there is a custom job script to launch. This script also accepts
arguments from the command-line can be useful for backfill.

### CODE REVIEW

**Comment by acmiyaguchi** [comment]
This PR is blocked by [telemetry-batch-view #316](https://github.com/mozilla/telemetry-batch-view/pull/316), the value of `--slack` in [bug 1407410](https://bugzilla.mozilla.org/show_bug.cgi?id=1407410), and a local run of Airflow.

A modified version of the run script was used to generate `retention_dev_v3`.

**Comment by acmiyaguchi** [comment]
The slack value is set to 4 to capture 99% of the activity for a particular session start date. This value can be cross-referenced with data from the [Telemetry Health Dashboard](https://sql.telemetry.mozilla.org/dashboard/telemetry-health).

**Comment by acmiyaguchi** [comment]
There's a weird error that's in the way of running this script successfully. It looks like some combination of the emr bootstrap script and the airflow script is spewing trash into the `--argument` variable.

```
+ MOZETL_COMMAND=retention
+ export MOZETL_RETENTION_START_DATE=20171001
+ MOZETL_RETENTION_START_DATE=20171001
+ export MOZETL_RETENTION_BUCKET=telemetry-test-bucket
+ MOZETL_RETENTION_BUCKET=telemetry-test-bucket
+ export MOZETL_RETENTION_PATH=intermediate_path
+ MOZETL_RETENTION_PATH=intermediate_path
+ MOZETL_GIT_BRANCH=master
+ MOZETL_GIT_PATH=https://github.com/mozilla/python_mozetl.git
+ TBV_GIT_PATH=https://github.com/mozilla/telemetry-batch-view.git
+ MOZETL_RUNNER_PATH=
+ MOZETL_EGG_PATH=
+ TBV_JAR_PATH=
+ main '""'
++ mktemp -d -t tmp.XXXXXXXXXX
+ workdir=/tmp/tmp.0QLmUFVpQ3
+ trap cleanup EXIT
+ MOZETL_RUNNER_PATH=/tmp/tmp.0QLmUFVpQ3/runner.py
+ cat
+ cd /tmp/tmp.0QLmUFVpQ3
+ prepare_mozetl
+ git clone https://github.com/mozilla/python_mozetl.git --branch master
Cloning into 'python_mozetl'...
+ cd python_mozetl
+ pip install .
You are using pip version 8.1.2, however version 9.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
+ python setup.py bdist_egg
zip_safe flag not set; analyzing archive contents...
++ ls dist/mozetl-0.1-py2.7.egg
+ local rel_egg_path=dist/mozetl-0.1-py2.7.egg
++ realpath dist/mozetl-0.1-py2.7.egg
+ MOZETL_EGG_PATH=/mnt/tmp/tmp.0QLmUFVpQ3/python_mozetl/dist/mozetl-0.1-py2.7.egg
+ cd /tmp/tmp.0QLmUFVpQ3
+ prepare_tbv
+ git clone https://github.com/mozilla/telemetry-batch-view.git --branch master
Cloning into 'telemetry-batch-view'...
+ cd telemetry-batch-view
+ sbt assembly
Picked up _JAVA_OPTIONS: -Djava.io.tmpdir=/mnt1/ -Xmx15000M -Xms1000M
Picked up _JAVA_OPTIONS: -Djava.io.tmpdir=/mnt1/ -Xmx15000M -Xms1000M
Exception in thread "main" java.nio.file.ProviderNotFoundException: Provider "jrt" not found
	at java.nio.file.FileSystems.getFileSystem(FileSystems.java:224)
	at io.github.retronym.java9rtexport.Export.main(Export.java:36)
Picked up _JAVA_OPTIONS: -Djava.io.tmpdir=/mnt1/ -Xmx15000M -Xms1000M
++ ls target/scala-2.11/telemetry-batch-view-1.1.jar
+ local rel_jar_path=target/scala-2.11/telemetry-batch-view-1.1.jar
++ realpath target/scala-2.11/telemetry-batch-view-1.1.jar
+ TBV_JAR_PATH=/mnt/tmp/tmp.0QLmUFVpQ3/telemetry-batch-view/target/scala-2.11/telemetry-batch-view-1.1.jar
+ cd /tmp/tmp.0QLmUFVpQ3
+ local 'start_date=""'
+ local 'end_date=""'
++ date -d '"" + 1 day' +%Y%m%d
date: invalid date â€˜"" + 1 dayâ€™
```

**Comment by acmiyaguchi** [comment]
A [successful run](https://us-west-2.console.aws.amazon.com/elasticmapreduce/home?region=us-west-2#cluster-details:j-D0FTYYUI8MVZ) on 5 instances ran 15 minutes over an hour, so I bumped the number of instances to 6. The results can be found under `s3://telemetry-test-bucket/retention/v1/start_date=20171002/
`.

@sunahsuh This is ready for review.

File: dags/main_summary.py [hunk 1]
```diff
@@ -194,6 +194,19 @@
     uri="https://raw.githubusercontent.com/mozilla/telemetry-airflow/master/jobs/heavy_users_view.sh",
     dag=dag)
 
+retention = EMRSparkOperator(task_id="retention",
```
Comment (sunahsuh) [inline dags/main_summary.py]: nit: put `task_id` on a newline for the hanging indent (pep-8)

File: dags/operators/emr_spark_operator.py [hunk 2]
```diff
@@ -91,22 +91,26 @@ def execute(self, context):
...
+            's3://{}/steps/airflow.sh'.format(
+                EMRSparkOperator.airflow_bucket
+            ),
+            '--job-name', self.job_name,
+            '--user', self.owner,
+            '--uri', self.uri,
+            '--data-bucket', self.data_bucket,
+            '--environment', self.environment
+        ]
+        if self.arguments:
```
Comment (sunahsuh) [inline dags/operators/emr_spark_operator.py]: Can you add a comment here referencing the bug this fixes?

**Comment by acmiyaguchi** [comment]
I agree the big bash script is not ideal, but it can be debugged and it handles backfill well.

The job shares a lot of history with `churn`, where I reused a large portion of utility functions and test harnesses. I'd rather use one of the established processes instead of creating a new edge case, but adding hyperloglog support to pyspark ended up being more dificult than I originally anticipated. My goal is to get `spark-hyperloglog` packaged in a way that is compatible with the `spark-2.2.0` pypi package. In the meanwhile, the scala job handles aggregating the HLL variable with some additional metrics.

I've filed and taken [bug 1408225](https://bugzilla.mozilla.org/show_bug.cgi?id=1408225) for simplifiying this process.
